### DISCUSSION TITLE
`type` vs `Type`  (pyupgrade issue)

### DISCUSSION THREAD

**Discussion by karolzlot** [microsoft/pyright#5391] [Q&A] [answered]
Context: 
"pyupgrade" precommit hook changes `Type` to `type`, however form pyright perspective those are not the same. I wonder if this is bug in pyright or bug in pyupgrade?

(BTW to disable pyupgrade conversion it is enough to use `from typing import Type as Type_`)

```python
from typing import cast, Any, Type


def get_from_db(key_name: str) -> Any:
    """Get a config value from the database"""
    value = ...
    return value


def get_value_A(key_name: str, type_: Type):
    # get value and check its type

    value = get_from_db(key_name)

    assert isinstance(value, type_)
    return cast(type_, value)


a1 = get_value_A("key1", str)
a2 = get_value_A("key2", int)

# above inlay hints as desired



def get_value_B(key_name: str, type_: type):
    # get value and check its type

    value = get_from_db(key_name)

    assert isinstance(value, type_)
    return cast(type_, value)


a1 = get_value_B("key1", str)
a2 = get_value_B("key2", int)

# above inlay hints not as desired
```

Note difference in inlay hints:

![image](https://github.com/microsoft/pyright/assets/11590960/44bbffb2-63dd-4b71-88cf-2db6a7620d8d)

**Comment by erictraut** [comment, marked as answer]
This is not a bug. The symbols `typing.Type` and `builtins.type` are different. The former is used only as a generic type and should always include a type argument, as in `Type[int]` or `Type[T]`. You're using it above without a type argument, so it is assumed to be `Type[Any]`. In contrast, the symbol `type` can be used either with or without a type argument. If used without a type argument, it refers to the class `type`.

The recommended way to write `get_value` is to use a type variable and make the function generic. This is the reason generics were added to the language.

```python
T = TypeVar("T")


def get_value(key_name: str, type_: type[T]) -> T:
    value = get_from_db(key_name)

    assert isinstance(value, type_)
    return value
```

**Reply by Azureblade3808** [reply]
I think we've missed some point.

We can have --
```python
from typing import cast

type_: type[str] = globals()[""]
value = cast(type_, None)
reveal_type(value)  # "str"
```
`str` can be replaced by whatever concrete type we can think of, and revealed type will also change accordingly. Except --
```python
from typing import Any, cast

type_: type[Any] = globals()[""]
value = cast(type_, None)
reveal_type(value)  # "type"???
```

I think the more correct behavior would be forbidding the usage of `cast(type_, None)` just like what Mypy does, or let the type of `value` be `Any` in the second sample.

**Reply by erictraut** [reply]
@Azureblade3808, yeah, that's a good point. Could you do me a favor and file a bug so this doesn't get lost?

**Reply by Azureblade3808** [reply]
Done in #5429  .
